src/fat.rs

```diff
@@ -0,0 +1,82 @@
+use anyhow::Context;
+use std::{collections::BTreeMap, fs, io, path::Path};
+
+use crate::KERNEL_FILE_NAME;
+
+pub fn create_fat_filesystem(
+    files: BTreeMap<&str, &Path>,
+    out_fat_path: &Path,
+) -> anyhow::Result<()> {
+    const MB: u64 = 1024 * 1024;
+
+    // calculate needed size
+    let mut needed_size = 0;
+    for path in files.values() {
+        let file_size = fs::metadata(path)
+            .with_context(|| format!("failed to read metadata of file `{}`", path.display()))?
+            .len();
+        needed_size += file_size;
+    }
+
+    // create new filesystem image file at the given path and set its length
+    let fat_file = fs::OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(&out_fat_path)
+        .unwrap();
+    let fat_size_padded_and_rounded = ((needed_size + 1024 * 64 - 1) / MB + 1) * MB;
+    fat_file.set_len(fat_size_padded_and_rounded).unwrap();
+
+    // choose a file system label
+    let mut label = *b"MY_RUST_OS!";
+    if let Some(path) = files.get(KERNEL_FILE_NAME) {
+        if let Some(name) = path.file_stem() {
+            let converted = name.to_string_lossy();
+            let name = converted.as_bytes();
+            let mut new_label = [0u8; 11];
+            let name = &name[..new_label.len()];
+            let slice = &mut new_label[..name.len()];
+            slice.copy_from_slice(name);
```

Comment:
This panics if the name is shorter than 11 bytes.

Reply:
Fixed in 960671c